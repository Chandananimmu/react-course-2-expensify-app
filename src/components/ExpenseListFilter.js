import React from "react";
import { connect } from "react-redux";
import { DateRangePicker } from "react-dates";
import {
  setFilterText,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate
} from "../actions/filters";

export class ExpenseListFilter extends React.Component {
  state = {
    calenderFocused: null
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };
  onFocusChange = calenderFocused => {
 
    this.setState(() => ({
      calenderFocused
    }));
  };
  onTextChange = e => {
    this.props.setFilterText({ text: e.target.value });
  };
  onSortChange = e => {
    e.target.value === "date"
      ? this.props.sortByDate()
      : this.props.sortByAmount();
  };
  render() {
    return (
      <div>
        <input
          type="text"
          value={this.props.filters.text}
          // onChange={e => {
          //   this.props.dispatch(setFilterText({ text: e.target.value }));
          // }}
          onChange={this.onTextChange}
        />
        <select
          value={this.props.filters.sortBy}
          // onChange={e => {
          //   this.props.dispatch(
          //     e.target.value === "date" ? sortByDate() : sortByAmount()
          //   );
          // }}
          onChange={this.onSortChange}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker 
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calenderFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
          showClearDates={true}
          
        />
      </div>
    );
  }
}
//const ExpenseListFilter = props =>
const mapStateToProps = state => {
  return {
    filters: state.filters
  };
};
const mapDispatchToProps = dispatch => ({
  setStartDate: startDate => {
    dispatch(setStartDate(startDate));
  },
  setEndDate: endDate => {
    dispatch(setEndDate(endDate));
  },
  setFilterText: text => {
    dispatch(setFilterText(text));
  },
  sortByDate: () => {
    dispatch(sortByDate());
  },
  sortByAmount: () => {
    dispatch(sortByAmount());
  }
});
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ExpenseListFilter);
