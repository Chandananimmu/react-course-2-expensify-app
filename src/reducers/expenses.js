const expensesReducerDefaultState = [];
export default (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case "ADD_EXPENSE":
      return [...state, action.expense];
    case "REMOVE_EXPENSE":
      const idd = action.id;
      return state.filter(({ id }) => id !== idd); //idd->the one wich v need to remove
    case "EDIT_EXPENSE":
      return state.map(expense => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.updates
          };
        } else return expense;
      });

    default:
      return state;
  }
};