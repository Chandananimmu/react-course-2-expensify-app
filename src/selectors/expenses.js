import moment from "moment";
const getVisibleExpenses = (
  expenses, {
    text,
    sortedBy,
    startDate,
    endDate
  }
) => {
  return expenses
    .filter(expense => {
      const createdAtMoment = moment(expense.createdAt)
      const startDateMatch =
        startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
      const EndDateMatch =
        endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
      //const startDateMatch=true;
      //const EndDateMatch=true
      //const t1=expense.description.toLowerCase();
      //const t2=text.toLowerCase();
      const textMatch = expense.description
        .toLowerCase()
        .includes(text.toLowerCase());

      return startDateMatch && EndDateMatch && textMatch;
    }).sort((a, b) => {
      if (sortedBy === 'date') {
        return a.createdAt < b.createdAt ? 1 : -1;
      } else
      if (sortedBy === 'amount') {
        return a.amount < b.amount ? 1 : -1;
      }
    });

};
export default getVisibleExpenses;