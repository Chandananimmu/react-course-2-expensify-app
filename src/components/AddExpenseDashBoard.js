import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { addExpense } from "../actions/expenses";
export class AddExpenseDashBoard extends React.Component {
  onSubmit = (expense) => {
    this.props.addExpense(expense);
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <p>This msg is from add dashboard </p>
        <ExpenseForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}
// const AddExpenseDashBoard = props => (
//   <div>
//     <p>This msg is from add dashboard </p>
//     <ExpenseForm
//       onSubmit={expense => {
//         //  console.log(expense);
//         props.onsubmit(expense);
//         props.history.push("/");
//       }}
//     />
//   </div>
// );
const mapDispatchToProps = (dispatch) => ({
  addExpense: (expense) => dispatch(addExpense(expense))
});
export default connect( undefined, mapDispatchToProps) (AddExpenseDashBoard);
 


