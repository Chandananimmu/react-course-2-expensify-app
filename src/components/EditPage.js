import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import { editExpense } from "../actions/expenses";
import { removeExpense } from "../actions/expenses";
export class EditPage extends React.Component {
  onSubmit =( expense) => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push("/");
  };
  onClick = () => {
    this.props.removeExpense({ id: this.props.expense.id });
    this.props.history.push("/");
  };
  render() {
    return (
      <div>
        <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
        <button onClick={this.onClick}>Remove</button>
      </div>
    );
  }
}
// const EditPage = props => (
//   <div>
//     <ExpenseForm
//     expense={props.expense}
//       onSubmit={(expense) => {
//        const id=props.expense.id
//         props.dispatch(editExpense(id,expense));
//         props.history.push("/")
//        // console.log("updated", expense);
//       }}
//     />
//      <button
//       onClick={() => {
//         const id=props.expense.id
//         props.dispatch(removeExpense({id}));
//         props.history.push("/")
//       }}
//     >
//       Remove
//     </button>
//   </div>
// );

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(
      expense => expense.id === props.match.params.id
    )
  };
};
const mapDispatchToProps = (dispatch, props) => ({
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  removeExpense: id => dispatch(removeExpense(id))
});
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditPage);
