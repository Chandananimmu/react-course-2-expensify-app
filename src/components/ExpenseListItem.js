import React from "react";
//import { connect } from "react-redux";
import { removeExpense } from "../actions/expenses";
import { Link } from "react-router-dom";
export const ExpenseListItem = ({ id, description, amount, createdAt }) => (
  <div>
    <Link to={`/edit/${id}`}>
      <h3> {description}</h3>
    </Link>

    <p> {amount}</p>
    <p>{createdAt}</p>
  </div>
);

// const mapDispatchFromProps = (push,dispatch) => ({
//  removeExpense: (id)=>dispatch(removeExpense({id})),

// })

//export default connect(null, mapDispatchFromProps)(ExpenseListItem);
export default ExpenseListItem;
