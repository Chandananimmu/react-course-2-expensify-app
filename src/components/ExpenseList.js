import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import getVisibleExpenses from "../selectors/expenses";
export const ExpenseList = props => (
  <div>
    <p>ExpenseList</p>
    {props.expenses.length === 0 ? (
      <p>No expenses</p>
    ) : (
      props.expenses.map(expense => {
        return <ExpenseListItem key={expense.id} {...expense} />;
      })
    )}
  </div>
);
const mapToStateProps = state => {
  return {
    expenses: getVisibleExpenses(state.expenses, state.filters)
  };
};

//const ConnnectedList = connect(mapToStateProps)(ExpenseList);
export default connect(mapToStateProps)(ExpenseList);
