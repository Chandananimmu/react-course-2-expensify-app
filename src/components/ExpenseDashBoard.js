import React from "react";
import ExpenseList from "./ExpenseList";
import ExpenseListFilter from "./ExpenseListFilter";
const ExpenseDashBoard = () => (
  <div>
    <ExpenseList />
    <ExpenseListFilter />
  </div>
);
export default ExpenseDashBoard;
