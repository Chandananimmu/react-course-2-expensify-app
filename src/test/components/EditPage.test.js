import React from "react";
import { EditPage } from "../../components/EditPage";
import { shallow } from "enzyme";
import expenses from "../fixtures/expenses";
//import { editExpense } from "../../actions/expenses";
let editExpense, history, wrapper, removeExpense;
beforeEach(() => {
  editExpense = jest.fn();
  removeExpense = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <EditPage
      editExpense={editExpense}
      removeExpense={removeExpense}
      history={history}
      expense={expenses[1]}
    />
  );
});
test("shud edit expense correctly", () => {
  expect(wrapper).toMatchSnapshot();
});
test("shud edit onSubmit", () => {
  wrapper.find("ExpenseForm").prop("onSubmit")(expenses[1]);
  expect(history.push).toHaveBeenLastCalledWith("/");
  expect(editExpense).toHaveBeenLastCalledWith(expenses[1].id, expenses[1]);
  expect(wrapper).toMatchSnapshot();
});
test("shud remove onClick", () => {
  wrapper.find("button").simulate("click");
  expect(history.push).toHaveBeenLastCalledWith("/");
  expect(removeExpense).toHaveBeenLastCalledWith({id:expenses[1].id});
  expect(wrapper).toMatchSnapshot();
});
