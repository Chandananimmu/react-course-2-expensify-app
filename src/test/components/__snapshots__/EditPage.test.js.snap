// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shud edit expense correctly 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 19500,
        "createdAt": -345600000,
        "description": "Rent",
        "id": "2",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;

exports[`shud edit onSubmit 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 19500,
        "createdAt": -345600000,
        "description": "Rent",
        "id": "2",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;

exports[`shud remove onClick 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 19500,
        "createdAt": -345600000,
        "description": "Rent",
        "id": "2",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
